samples/smp/pktqueue/src/main.c

```diff
@@ -0,0 +1,198 @@
+int current_queue;
+
+/* Array of packet header descriptors */
+struct phdr_desc descriptors[QUEUE_NUM][SIZE_OF_QUEUE];
+
+/* Arrays of receiver and sender queues */
+struct phdr_desc_queue sender[QUEUE_NUM], receiver[QUEUE_NUM];
+
+/* Array of packet headers */
+uint8_t headers[QUEUE_NUM][SIZE_OF_QUEUE][SIZE_OF_HEADER];
+
+static K_THREAD_STACK_ARRAY_DEFINE(tstack, THREADS_NUM*QUEUE_NUM, STACK_SIZE);
+static K_THREAD_STACK_ARRAY_DEFINE(qstack, QUEUE_NUM, STACK_SIZE);
+
+K_MUTEX_DEFINE(fetch_queue_mtx);
+
+/* Function for initializing "sender" packet header queue */
+void init_datagram_queue(struct phdr_desc_queue *queue, int queue_num)
+{
+	queue->head = descriptors[queue_num];
+
+	for (int i = 0; i < SIZE_OF_QUEUE; i++) {
+		queue->tail = &descriptors[queue_num][i];
+		descriptors[queue_num][i].ptr = (uint8_t *)&headers[queue_num][i];
+		/* Fill packet header with random values */
+		for (int j = 0; j < SIZE_OF_HEADER; j++) {
+			/* leave crc field zeroed */
+			if (j < 10 || j > 11)
+				descriptors[queue_num][i].ptr[j] = (uint8_t)sys_rand32_get();
+			else
+				descriptors[queue_num][i].ptr[j] = 0;
+		}
+		/* Compute crc for further comparisson */
+		uint16_t crc;
+
+		crc = crc16(descriptors[queue_num][i].ptr, SIZE_OF_HEADER,
+					POLYNOMIAL, 0, 0);
+
+		/* Save crc value in header[10-11] field */
+		descriptors[queue_num][i].ptr[10] = (uint8_t)(crc >> 8);
+		descriptors[queue_num][i].ptr[11] = (uint8_t)(crc);
+		queue->count++;
+		descriptors[queue_num][i].next = &descriptors[queue_num][i+1];
+	}
+}
+
+/* Thread takes packet from "sender" queue and puts it to "receiver" queue.
+ * Each queue can be accessed only by one thread in a time. */
+void test_thread(void *arg1, void *arg2, void *arg3)
+{
+	struct phdr_desc_queue *sender_queue = (struct phdr_desc_queue *)arg1;
+	struct phdr_desc_queue *receiver_queue = (struct phdr_desc_queue *)arg2;
+	struct phdr_desc *qin_ptr = NULL;
+	int queue_num = *(int *)arg3;
+
+	/* Fetching one queue */
+
+	uint16_t crc, crc_orig;
+
+	qin_ptr = phdr_desc_dequeue(sender_queue, &sender_queue_mtx[queue_num]);
+	while (qin_ptr != NULL) {
+		/* Store original crc value from header */
+		crc_orig  =  qin_ptr->ptr[10] << 8;
+		crc_orig |=   qin_ptr->ptr[11];
+
+		/* Crc field should be zero before crc calculation */
+		qin_ptr->ptr[10] = 0;
+		qin_ptr->ptr[11] = 0;
+		crc = crc16(qin_ptr->ptr, SIZE_OF_HEADER, POLYNOMIAL, 0, 0);
+
+		/* Compare computed crc with crc from phdr_desc->crc */
+		if (crc == crc_orig) {
+			phdr_desc_enqueue(receiver_queue, qin_ptr,
+						 &receiver_queue_mtx[queue_num]);
+		}
+		/* Take next element from "sender queue" */
+		qin_ptr = phdr_desc_dequeue(sender_queue,
+						&sender_queue_mtx[queue_num]);
+	}
+}
+
+/* Thread that processes one pair of sender/receiver queue */
+void queue_thread(void *arg1, void *arg2, void *arg3)
+{
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	int queue_num;
+	/* Fetching one queue */
```

Comment:
I'd add a newline before comment for better visual separation of the logical block.